dataframe.py

```diff
@@ -0,0 +1,269 @@
+                self.category_type == other.category_type and
+                self.ordered == other.ordered)
+
+    def to_string(self):
+        return ("categorical(indices={}, categories={}, ordered={})"
+                .format(str(self.index_type), str(self.category_type),
+                        self.ordered))
+
+
+# ----------------------------------------------------------------------
+# Classes representing a column in a DataFrame
+
+
+class Column(ABC):
+
+    @property
+    @abstractmethod
+    def name(self) -> Hashable:
+        pass
+
+    @property
+    @abstractmethod
+    def type(self) -> DataType:
+        """
+        Return the logical type of each column cell value
+        """
+        pass
+
+    @property
+    def attrs(self) -> Mapping:
+        """
+        Metadata for this column. Default implementation returns empty dict
+        """
+        return {}
+
+    def to_numpy(self):
+        """
+        Access column's data as a NumPy array. Recommended to return a view if
+        able but not required
+        """
+        raise NotImplementedError("Conversion to NumPy not available")
+
+    def to_arrow(self, **kwargs):
+        """
+        Access column's data in the Apache Arrow format as pyarrow.Array or
+        ChunkedArray. Recommended to return a view if able but not required
+        """
+        raise NotImplementedError("Conversion to Arrow not available")
+
+
+# ----------------------------------------------------------------------
+# DataFrame: the main public API
+
+
+class DataFrame(ABC):
+    """
+    An abstract data frame base class.
+
+    A "data frame" represents an ordered collection of named columns. A
+    column's "name" is permitted to be any hashable Python value, but strings
+    are common. Names are not required to be unique. Columns may be accessed by
+    name (when the name is unique) or by position.
+    """
+
+    def __dataframe__(self):
+        """
+        Idempotence of data frame protocol
+        """
+        return self
+
+    @property
+    @abstractmethod
+    def num_columns(self):
+        """
+        Return the number of columns in the DataFrame
+        """
+        pass
+
+    @property
+    @abstractmethod
+    def num_rows(self):
+        """
+        Return the number of rows in the DataFrame (if known)
+        """
+        pass
+
+    @property
+    @abstractmethod
+    def column_names(self) -> Sequence[Any]:
+        """
+        Return the column names as a materialized sequence
+        """
+        pass
+
+    @property
+    def row_names(self) -> Sequence[Any]:
+        """
+        Return the row names (if any) as a materialized sequence. It is not
+        necessary to implement this method
+        """
+        raise NotImplementedError("This DataFrame has no row names")
+
+    @abstractmethod
+    def get_column(self, i: int) -> Column:
+        """
+        Return the column at the indicated position
+        """
+        pass
```

Comment:
From a Python dev perspective (who would like to use this functionality) the get/select methods seem a bit awkward.

If we instead extend the `columns(self) -> ColumnCollection:` model we can have an api like:
```python
In [79]: df = DataFrame('A', 'B', 'C')

In [80]: df
Out[80]: DataFrame(['A', 'B', 'C'])

In [81]: df.columns
Out[81]: ColumnCollection(['A', 'B', 'C'])

In [82]: df.columns.idx(0)
Out[82]: Column(name='A')

In [83]: df.columns.idx(0, 2)
Out[83]: DataFrame(['A', 'C'])

In [84]: df.columns['A']
Out[84]: Column(name='A')

In [85]: df.columns['A','C']
Out[85]: DataFrame(['A', 'C'])
```
...which, at least to me, seems a lot more natural.

<details>
<summary><tt><b>Code:</b></tt></summary>

```python

from typing import List
from dataclasses import dataclass


@dataclass
class Column(object):
    name: str


class ColumnCollection(object):
    _columns: List[Column]
    
    def __init__(self, *columns: Column):
        self._columns = list(columns)
    
    @property
    def names(self) -> List[str]:
        return [col.name for col in self._columns]

    def idx(self, *indices: int) -> ColumnCollection:
        columns = [self._columns[idx] for idx in indices]
        if len(columns) == 1:
            return columns[0]
        return DataFrame(
            *(col.name for col in columns)
        )
    
    def __getitem__(self, names: str) -> Union[Column, ColumnCollection]:
        columns = dict(zip(self.names, self._columns))
        if len(names) == 1:
            return columns[names[0]]
        return DataFrame(*names)
    
    def __len__(self):
        return len(self._columns)

    def __iter__(self):
        for col in self._columns:
            yield col
        
    def __repr__(self):
        columns = "', '".join(self.names)
        return f"ColumnCollection(['{columns}'])"


class DataFrame(object):
    _columns: ColumnCollection
    
    def __init__(self, *columns):
        self._columns = ColumnCollection(
            *(Column(col) for col in columns)
        )
        
    def __repr__(self):
        columns = "', '".join(self.columns.names)
        return f"DataFrame(['{columns}'])"   

    @property
    def columns(self):
        return self._columns
```
<details>

Reply:
> From a Python dev perspective (who would like to use this functionality) the get/select methods seem a bit awkward.

I strongly feel that syntactic sugar / conveniences (which get non-objective very quickly) belong at a higher level than this API, which is about library developers implementing unambiguous data export APIs.

Reply:
I'm interested in this as a potential way to write analytics functions which are DataFrame implementation agnostic. i.e. instead of coding to the pandas API I would code to this API hence its usage would be front-and-centre in my code in the same way that pandas now is. Maybe this isn't supposed to cater to that end-user concern though?

Reply:
It's not. This was discussed at length in https://discuss.ossdata.org/t/a-dataframe-protocol-for-the-pydata-ecosystem/267